Applications/Terminal/TerminalWidget.cpp

```diff
@@ -225,21 +225,57 @@ void TerminalWidget::paint_event(GPaintEvent& event)
             painter.fill_rect(row_rect, Color::Red);
         else if (has_only_one_background_color)
             painter.fill_rect(row_rect, lookup_color(line.attributes[0].background_color).with_alpha(m_opacity));
-        for (u16 column = 0; column < m_terminal.columns(); ++column) {
-            char ch = line.characters[column];
-            bool should_reverse_fill_for_cursor_or_selection = (m_cursor_blink_state && m_in_active_window && row == row_with_cursor && column == m_terminal.cursor_column())
-                || selection_contains({ row, column });
-            auto& attribute = line.attributes[column];
-            auto character_rect = glyph_rect(row, column);
-            if (!has_only_one_background_color || should_reverse_fill_for_cursor_or_selection) {
-                auto cell_rect = character_rect.inflated(0, m_line_spacing);
-                painter.fill_rect(cell_rect, lookup_color(should_reverse_fill_for_cursor_or_selection ? attribute.foreground_color : attribute.background_color).with_alpha(m_opacity));
+
+
+        // The terminal insists on thinking characters and
```

Comment:
We should probably refactor the terminal `Line` class to conceptually be a series of `u32` codepoints.
It could have two storage modes, a compact 1-byte-per-cell mode where it's just ASCII text, and a full 4-bytes-per-cell mode for arbitrary codepoints.
What do you think?

Reply:
We could do that (and VTE supports it somehow, so it should be possible), but the complication is I imagine a lot of code in the terminal ecosystem (text editors and such) assumes the *grid of characters which are the same as bytes* model. Serenity's `ls` does something like that to align file names into columns, for instance. We can of course change `ls`, but I assume lots of other programs would break if outputting n bytes didn't mean moving n columns to the right, etc.

I'm not saying we shouldn't try to do this; but it's complicated.